sisl/io/siesta/eig.py

```diff
@@ -140,29 +144,177 @@ def opts(*args):
+                        key, val = value.split("=")
+                        if key.lower() in ("de", "e"):
+                            try:
+                                dE = units(val, "eV")
+                            except Exception:
+                                dE = float(val)
+                        elif key.lower() in ("kt", "t"):
+                            try:
+                                kT = units(val, "eV")
+                            except Exception:
+                                kT = float(val)
+                        elif key.lower().startswith("dist"):
+                            distribution = val
+                        else:
+                            raise ValueError(f"Unknown key: {key}, should be one of [dE, kT, dist]")
+
+                    elif i == 0:
+                        try:
+                            dE = float(value)
+                        except Exception:
+                            # *must* be a str?
+                            distribution = value
+                    elif i == 1:
+                        try:
+                            kT = float(value)
+                        except Exception:
+                            # *must* be a str?
+                            distribution = value
+                    elif i == 2:
+                        try:
+                            distribution = value
+                        except Exception:
+                            raise ValueError(f"Unknown value for distribution: {value}")
+                    else:
+                        raise ValueError(f"Too many values passed? Unknown value {value}?")
+
+                # store for next invocation
+                ns._dos_args[0] = dE
+                ns._dos_args[1] = kT 
+                ns._dos_args[2] = distribution
+
+                # Now create the final distribution
+                distribution = get_distribution(distribution, smearing=kT)
+                
+                if ns._Emap is None:
+                    # We will plot the DOS in the entire energy window
+                    ns._Emap = [ns._eigs.min() - kT * 4, ns._eigs.max() + kT * 4]
+
+                # Now we are ready to process
+                E = np.arange(ns._Emap[0], ns._Emap[1], dE)
+                n_data = len(ns._data_header)
+                same_E = False
+                if n_data > 0:
+                    for i in range(n_data):
+                        header = ns._data_header[-i]
+                        if header == "Energy":
+                            try:
+                                same_E = np.allclose(E, ns._data[-i])
+                            except Exception:
+                                same_E = False
+
+                if not same_E:
+                    ns._data_header.append("Energy")
+                    ns._data.append(E)
+
+                def calc_dos(E, eig, w):
+                    nonlocal distribution
```

Comment:
Is there any specific reason, why it is beneficial to choose `nonlocal` over passing as a argumetn?

Reply:
No, none at all in this case... :)